Add support for Python 3.13

jgraham opened the pull request:
Notably this involves vendoring a copy of the removed cgi module from stdlib.

jgraham committed:
Import fork of Python cgi library into wptserve

This code has been dropped from Python 3.13, but we depend on
FieldStorage as part of the request API. The easiest option here is to
directly import the code that we are already using (the last version
on the 3.12 branch, but with the deprecation warning removed), along
with a copy of the PSF License which covers the cgi module code.

jgraham committed:
Use vendored cgi lib in wptserve

jgraham committed:
Update third party deps for Python 3.13 support

jgraham committed:
Fix lock in QueueHandler class to be a dummy implementation of the Lock API

This is required since the logging module started using the lock as a context manager,
and calling the acquire and release methods.

jgraham committed:
Update getting context manager when running serve directly

Remove code only needed for Python < 3.4 and ensure that we always use the spawn
form of multiprocessing to avoid problems with fork and locks.

jgraham committed:
Update tox configurations to include Python 3.13

jgraham committed:
Update Azure configuration for Python 3.13

jgraham committed:
Update taskcluster configuration for Python 3.13

jgraham committed:
Exclude cgi code from flake8

gsnedders reviewed (commented):
I wonder if we should also vendor the cgi tests, for the sake of any changes we make to it?

jgraham committed:
Ignore errors in imported cgi module

Just adding this to the excludes doesn't work, because we still check the imported
code.

jgraham commented on tools/wptrunner/wptrunner/wptlogging.py:
```diff
@@ -77,7 +77,7 @@ def __init__(self, queue, level=logging.NOTSET):
     def createLock(self):
         # The queue provides its own locking
-        self.lock = None
+        self.lock = NullLock()
```
Unfortunately it very much does (I spent a while figuring out why we were getting a failure to use `None` as a context manager in the logging module, without any traceback pointing to this code).

https://github.com/python/cpython/issues/109461 seems to be the relevant upstream change.

jgraham committed:
Add cgi tests

Lightly modified from the version in Python 3.12 as follows:
* Inlined check__all__ function
* Converted the use of ignore_warnings helper to use warning module directly

jgraham committed:
Update Pillow for Python>=3.9

jgraham committed:
Remove dep on typed-ast

This is no longer support in Python 3.13, and no longer required for Python>=3.8

jgraham committed:
Fix cgi tests to be compatible with behaviour of older Python versions

gsnedders commented on tools/wptrunner/wptrunner/wptlogging.py:
```diff
@@ -90,6 +90,21 @@ def emit(self, record):
 
 
+
+class NullLock:
```
Given this is brand new code, can we please actually add types for all of this? This should, I think, match: https://github.com/python/typeshed/blob/614e9499f145e9dccade16d906247276f1d5e6aa/stdlib/_thread.pyi#L16-L20

Can we also add a comment saying that this implements the interface of `threading.RLock`, and probably rename this to `NullRLock`, to clarify this doesn't match `threading.Lock`?

jgraham committed:
Add types to NullLock and rename to NullRLock

gsnedders reviewed (changes requested):
https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=131812&view=logs&j=83e18e9b-478b-5a9d-ba56-9f00eadc4c2c&t=66b9c650-1fb4-5319-b61a-668ad58b4535 is still running `tox -c tools/ -e py311 -- --junitxml=results.xml`

Ms2ger commented on tools/wave/tox.ini:
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py38,py39,py310,py311,py312
+envlist = py38,py39,py310,py311,py312,py312
```
```suggestion
envlist = py38,py39,py310,py311,py312,py313
```

Ms2ger commented on tools/wpt/tox.ini:
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py38,py39,py310,py311,py312
+envlist = py38,py39,py310,py311,py312,py312
```
```suggestion
envlist = py38,py39,py310,py311,py312,py313
```

Commit:
Update assert_png to check for magic bytes directly

This avoids the use of imghdr which was removed in Ptyhon 3.13
